charger/ocpp/cp_core.go

```diff
@@ -111,6 +111,17 @@ func (cp *CP) StopTransaction(request *core.StopTransactionRequest) (*core.StopT
 
 	conn := cp.connectorByTransactionID(request.TransactionId)
 	if conn == nil {
+		if request.Timestamp != nil {
```

Comment:
What is the intention here- what makes a `StopTransaction` "old" here?

Reply:
The box sends a Stop for a transaction (in my cases it was always related to vehicle disconnects) that was kicked of by evcc hours or days before but never acknowledged by it. It comes in just after startup ( a little extra verbose here):

2024-05-20T12:57:11.304048+02:00 raspberrypi evcc[1863340]: [main  ] INFO 2024/05/20 12:57:11 starting ui and api at :7070
2024-05-20T12:57:11.474187+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] DEBUG 2024/05/20 12:57:11 waiting for chargepoint: 5m0s
2024-05-20T12:57:11.929173+02:00 raspberrypi evcc[1863340]: [ocpp  ] DEBUG 2024/05/20 12:57:11 charge point connected: 30J-HZH-5R4
2024-05-20T12:57:11.929342+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] DEBUG 2024/05/20 12:57:11 Franz: NewOCPP after wait %!t(<-chan struct {}=0x4001428120)
2024-05-20T12:57:11.929549+02:00 raspberrypi evcc[1863340]: [ocpp  ] TRACE 2024/05/20 12:57:11 sent JSON message to 30J-HZH-5R4: [2,"170851936","TriggerMessage",{"requestedMessage":"BootNotification","c
onnectorId":1}]
2024-05-20T12:57:11.957548+02:00 raspberrypi evcc[1863340]: [ocpp  ] TRACE 2024/05/20 12:57:11 received JSON message from 30J-HZH-5R4: [2,"1000163","StopTransaction",{"meterStop":851389,"timestamp":"202
4-05-20T09:26:55.039Z","transactionId":1,"reason":"EVDisconnected"}]
2024-05-20T12:57:11.958196+02:00 raspberrypi evcc[1863340]: [ocpp  ] TRACE 2024/05/20 12:57:11 Franz: OnStopTransaction received for 30J-HZH-5R4 request: %!w(*core.StopTransactionRequest=&{ 851389 0x400
0dda5e8 1 EVDisconnected []}) err: %!w(<nil>)
2024-05-20T12:57:11.958342+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] TRACE 2024/05/20 12:57:11 Franz: cp.StopTransaction received with request: %!w(*core.StopTransactionRequest=&{ 851389 0x4000dd
a5e8 1 EVDisconnected []})
2024-05-20T12:57:11.958754+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] TRACE 2024/05/20 12:57:11 Franz: cp.StopTransaction received with Transaction: %!w(int=1) conn: %!w(*ocpp.Connector=<nil>)
2024-05-20T12:57:11.958833+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] TRACE 2024/05/20 12:57:11 Franz: cp.StopTransaction received with Timestamp: %!w(*types.DateTime=&{{39000000 63851794015 <nil>
}}) timestamp: %!w(time.Time={39000000 63851794015 <nil>}) expiry: %!w(time.Duration=3600000000000)
2024-05-20T12:57:11.959037+02:00 raspberrypi evcc[1863340]: [30J-HZH-5R4-1] TRACE 2024/05/20 12:57:11 Franz: expired StopTransaction received and rejecting it
2024-05-20T12:57:11.959205+02:00 raspberrypi evcc[1863340]: [ocpp  ] TRACE 2024/05/20 12:57:11 sent JSON message to 30J-HZH-5R4: [3,"1000163",{"idTagInfo":{"status":"Expired"}}]

I for this example I manually created such a situation by doing os.Exit in OnStopTransaction before. But last Friday I received the same transaction all day:

2024-05-17T07:47:11.569415+02:00 raspberrypi evcc[1813951]: [ocpp  ] TRACE 2024/05/17 07:47:11 received JSON message from 30J-HZH-5R4: [2,"1001026","StopTransaction",{"meterStop":808626,"timestamp":"2024-05-17T05:14:14.470Z","transactionId":1,"reason":"EVDisconnected"}]
..
2024-05-17T07:51:33.721534+02:00 raspberrypi evcc[1813951]: [ocpp  ] TRACE 2024/05/17 07:51:33 received JSON message from 30J-HZH-5R4: [2,"1001026","StopTransaction",{"meterStop":808626,"timestamp":"2024-05-17T05:14:14.470Z","transactionId":1,"reason":"EVDisconnected"}]
..
2024-05-17T07:54:27.723985+02:00 raspberrypi evcc[1813981]: [ocpp  ] TRACE 2024/05/17 07:54:27 received JSON message from 30J-HZH-5R4: [2,"1001026","StopTransaction",G{"meterStop":808626,"timestamp":"2024-05-17T05:14:14.470Z","transactionId":1,"reason":"EVDisconnected"}]

Which lead to evcc not being able to start:
2024-05-17T10:07:26.232428+02:00 raspberrypi evcc[1814798]: [30J-HZH-5R4-1] DEBUG 2024/05/17 10:07:26 waiting for chargepoint: 5m0s
2024-05-17T10:07:30.351122+02:00 raspberrypi evcc[1814798]: [ocpp  ] DEBUG 2024/05/17 10:07:30 charge point connected: 30J-HZH-5R4
2024-05-17T10:07:30.351631+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:30 sent JSON message to 30J-HZH-5R4: [2,"1189722373","TriggerMessage",{"requestedMessage":"BootNotification","
connectorId":1}]
2024-05-17T10:07:30.365303+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:30 received JSON message from 30J-HZH-5R4: [2,"1001026","StopTransaction",{"meterStop":808626,"timestamp":"202
4-05-17T05:14:14.470Z","transactionId":1,"reason":"EVDisconnected"}]
2024-05-17T10:07:30.371440+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:30 received JSON message from 30J-HZH-5R4: [3,"1189722373",{"status":"Accepted"}]
2024-05-17T10:07:30.371936+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:30 TriggerMessage BootNotification for 30J-HZH-5R4: Accepted
2024-05-17T10:07:31.353581+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:31 sent JSON message to 30J-HZH-5R4: [2,"1287729144","GetConfiguration",{}]
2024-05-17T10:07:31.369897+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:31 received JSON message from 30J-HZH-5R4: [3,"1287729144",{"configurat..
..
2024-05-17T10:07:31.391519+02:00 raspberrypi evcc[1814798]: [30J-HZH-5R4-1] DEBUG 2024/05/17 10:07:31 enabling meter watchdog
2024-05-17T10:07:31.392049+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:31 sent JSON message to 30J-HZH-5R4: [2,"367761679","TriggerMessage",{"requestedMessage":"MeterValues","connectorId":1}]
2024-05-17T10:07:31.392623+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:31 received JSON message from 30J-HZH-5R4: [3,"367761679",{"status":"Accepted"}]
2024-05-17T10:07:31.392917+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:07:31 TriggerMessage MeterValues for 30J-HZH-5R4: Accepted
2024-05-17T10:08:31.378147+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:08:31 sent JSON message to 30J-HZH-5R4: [2,"1343343025","TriggerMessage",{"requestedMessage":"StatusNotification","connectorId":1}]
2024-05-17T10:08:31.395958+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:08:31 received JSON message from 30J-HZH-5R4: [3,"1343343025",{"status":"Accepted"}]
2024-05-17T10:08:31.396334+02:00 raspberrypi evcc[1814798]: [ocpp  ] TRACE 2024/05/17 10:08:31 TriggerMessage StatusNotification for 30J-HZH-5R4: Accepted
2024-05-17T10:09:31.378601+02:00 raspberrypi evcc[1814798]: [main  ] FATAL 2024/05/17 10:09:31 cannot create charger 'enercab': cannot create charger type 'template': cannot create charger type 'ocpp': timeout
2024-05-17T10:09:31.379127+02:00 raspberrypi evcc[1814798]: [main  ] FATAL 2024/05/17 10:09:31 will attempt restart in: 5m0s

Since conn is still nil at the time, checking for the Timestamp is actually a little more than necessary, because it has to come from something older at startup anyway. Unplugging the box, hard or soft resets, and last time I also turned ocpp on and off did not resolve this - all things that can not be expected to be done in normal use. Only starting iobroker-ocpp which consumed the stoptransaction right away did, so I added such a feature as well.

Reply:
So the point here is that evcc should simply authorise any StopTransaction it initially receives, i.e. while conn is still nil? But the box is happy with the `types.AuthorizationStatusExpired` status?

Reply:
yes it is happy with it - I would like to try it with AuthorizationStatusAccepted but I can't charge right now to test it.

Reply:
it works with sending an accept message as well, so I changed it to sending an accept, so that the box can close that task for itself.